kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala

```diff
@@ -53,77 +55,155 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+              logger.warn(s"the tail of ${current.meta.attribs.path} is to be followed, but it shrunk")
+              (current.meta.inspectedNow().modifiedNow(), EmptyFileBody)
             }
           } else {
-            // the file shrunk or didn't grow
-            logger.warn(s"the tail of ${current.meta.attribs.path} is to be followed, but it shrunk")
-            (current.meta.inspectedNow().modifiedNow(), EmptyFileBody)
+            // !tail: we're not tailing but dumping the entire file on change
+            logger.info(s"dump entire ${current.meta.attribs.path}")
+            (current.meta.inspectedNow().modifiedNow(), FileBody(current.body, 0))
           }
-        } else {
-          // !tail: we're not tailing but dumping the entire file on change
+        case Some(_) =>
+          // file didn't change
+          logger.info(s"fetched ${current.meta.attribs.path}, it was known before and it didn't change")
+          (current.meta.inspectedNow(), EmptyFileBody)
+        case None =>
+          // file is new
+          logger.info(s"fetched ${current.meta.attribs.path}, bytes= ${current.meta.attribs.size} , wasn't known before")
           logger.info(s"dump entire ${current.meta.attribs.path}")
           (current.meta.inspectedNow().modifiedNow(), FileBody(current.body, 0))
-        }
-      case Some(_) =>
-        // file didn't change
-        logger.info(s"fetched ${current.meta.attribs.path}, it was known before and it didn't change")
-        (current.meta.inspectedNow(), EmptyFileBody)
-      case None =>
-        // file is new
-        logger.info(s"fetched ${current.meta.attribs.path}, wasn't known before")
-        logger.info(s"dump entire ${current.meta.attribs.path}")
-        (current.meta.inspectedNow().modifiedNow(), FileBody(current.body, 0))
+      }
+    } else {
+      prevFetch match {
+        case Some(previously) if previously.attribs.timestamp != current.meta.attribs.timestamp =>
+          logger.info(s"distant file was modified since last fetch, previously timestamp=${previously.attribs.timestamp}, current timestamp=${current.meta.attribs.timestamp}")
+          logger.info(s"need to read from offset 0")
+          (current.meta.inspectedNow().offset(0), FileBody(current.body, 0))
+        case Some(previously) =>
+          logger.info(s"fetched ${current.meta.attribs.path}, bytes read= ${current.body.length} , from offset=${previously.offset}")
+          val nextOffsetToReadFrom = previously.offset + current.body.length
+          (current.meta.inspectedNow().offset(nextOffsetToReadFrom), FileBody(current.body, nextOffsetToReadFrom))
+        case None =>
+          // chunk is new
+          logger.info(s"fetched ${current.meta.attribs.path}, bytes read= ${current.body.length} , from offset=0")
+          val nextOffsetToReadFrom = current.body.length
+          (current.meta.inspectedNow().offset(nextOffsetToReadFrom), FileBody(current.body, nextOffsetToReadFrom))
+      }
```

Comment:
I'm having a very hard time following the ifs-in-ifs-in-matches. Could a suggest breaking up in function so the we can read separately the control flow logic and the actions taken in each branch ?